<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        
        
        <link rel="shortcut icon" href="../img/favicon.ico">

        <title>Changelog - SARS2seq</title>

        <link href="../css/bootstrap.min.css" rel="stylesheet">
        <link href="../css/font-awesome.min.css" rel="stylesheet">
        <link rel="stylesheet" href="../css/highlight.min.css">
        <link href="../css/base.css" rel="stylesheet">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>

            <!-- Main title -->
            <a class="navbar-brand" href="..">
            
            SARS2seq
            </a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            <!-- Main navigation -->
            <ul class="nav navbar-nav">
            
            
            
                
            
            
            
            
            
            </ul>

            <!-- Search, Navigation and Repo links -->
            <ul class="nav navbar-nav navbar-right">
            
            
            
            
            
                
                <li class="active">
                    <a href="./">
                      Changelog
                    </a>
                </li>
                
            
            
            
              
                
                <li>
                    <a href="https://github.com/RIVM-bioinformatics/SARS2seq">
                        
                            <i class="fa fa-code-fork"></i>
                        
                        Github
                    </a>
                </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            <div class="col-md-3">
  
<div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#changelog">Changelog</a></li>
        
            <li><a href="#068-2024-07-03">0.6.8 (2024-07-03)</a></li>
        
            <li><a href="#067-2024-04-04">0.6.7 (2024-04-04)</a></li>
        
            <li><a href="#066-2023-03-01">0.6.6 (2023-03-01)</a></li>
        
            <li><a href="#065-2023-01-17">0.6.5 (2023-01-17)</a></li>
        
            <li><a href="#064-2022-10-10">0.6.4 (2022-10-10)</a></li>
        
            <li><a href="#063-2022-10-10">0.6.3 (2022-10-10)</a></li>
        
            <li><a href="#062-2022-08-23">0.6.2 (2022-08-23)</a></li>
        
            <li><a href="#061-2022-08-15">0.6.1 (2022-08-15)</a></li>
        
            <li><a href="#060-2022-04-26">0.6.0 (2022-04-26)</a></li>
        
            <li><a href="#050-2022-01-27">0.5.0 (2022-01-27)</a></li>
        
            <li><a href="#040-2021-11-25">0.4.0 (2021-11-25)</a></li>
        
            <li><a href="#030-2021-11-09">0.3.0 (2021-11-09)</a></li>
        
    
    </ul>
</div>
  
</div>
            <div class="col-md-9" role="main">

<h1 id="changelog">Changelog</h1>
<h2 id="068-2024-07-03"><a href="https://github.com/RIVM-bioinformatics/SARS2seq/compare/v0.6.7...v0.6.8">0.6.8</a> (2024-07-03)</h2>
<h3 id="dependencies">Dependencies</h3>
<ul>
<li>Update conda environment files to remove the 'intel' channel (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/4572252e6114937ce05ffd2e8fae60602cc0582a">4572252</a>)</li>
</ul>
<h3 id="documentation">Documentation</h3>
<ul>
<li>update documentation page accordingly (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/4572252e6114937ce05ffd2e8fae60602cc0582a">4572252</a>)</li>
</ul>
<h2 id="067-2024-04-04"><a href="https://github.com/RIVM-bioinformatics/SARS2seq/compare/v0.6.6...v0.6.7">0.6.7</a> (2024-04-04)</h2>
<h3 id="bug-fixes">Bug Fixes</h3>
<ul>
<li>properly handle empty index files (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/0ee4d74e3b2530aadcd83968384ed72b1fb00ada">0ee4d74</a>)</li>
</ul>
<h3 id="dependencies_1">Dependencies</h3>
<ul>
<li>explicitly state the nodefaults channel in all environments (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/749ad4a544ef8616e2a36688fecf138ff868aea9">749ad4a</a>)</li>
<li>remove anaconda and defaults channels from conda recipes (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/e201894d0ddb43e441a4eec93087541ce2b5e98a">e201894</a>)</li>
<li>solve minor installation hassles for the main environment (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/749ad4a544ef8616e2a36688fecf138ff868aea9">749ad4a</a>)</li>
</ul>
<h2 id="066-2023-03-01"><a href="https://github.com/RIVM-bioinformatics/SARS2seq/compare/v0.6.5...v0.6.6">0.6.6</a> (2023-03-01)</h2>
<h3 id="dependencies_2">Dependencies</h3>
<ul>
<li>Update AmpliGone to v1.2.1 (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/67706ffaca501c228a6a09e92753ffb8693df4dc">67706ff</a>)</li>
</ul>
<h2 id="065-2023-01-17"><a href="https://github.com/RIVM-bioinformatics/SARS2seq/compare/v0.6.4...v0.6.5">0.6.5</a> (2023-01-17)</h2>
<h3 id="code-refactoring">Code Refactoring</h3>
<ul>
<li>re-balance threads to improve parallelization in local execution mode (use at most n threads for certain steps) (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/a8bccade7f96861e3563dd915d0217fe7d61bd2b">a8bccad</a>)</li>
</ul>
<h2 id="064-2022-10-10"><a href="https://github.com/RIVM-bioinformatics/SARS2seq/compare/v0.6.3...v0.6.4">0.6.4</a> (2022-10-10)</h2>
<h3 id="dependencies_3">Dependencies</h3>
<ul>
<li>set tabulate version to 0.8.10 in Typing environment as workaround (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/a3918648127d36ae89dfedf33f37a7ea89fa1829">a391864</a>)</li>
</ul>
<h2 id="063-2022-10-10"><a href="https://github.com/RIVM-bioinformatics/SARS2seq/compare/v0.6.2...v0.6.3">0.6.3</a> (2022-10-10)</h2>
<h3 id="dependencies_4">Dependencies</h3>
<ul>
<li>pin python-tabulate to v0.8.10 as a workaround for snakemake (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/3766878de489da3d78021392e5b61fd8879322d0">3766878</a>)</li>
</ul>
<h2 id="062-2022-08-23"><a href="https://github.com/RIVM-bioinformatics/SARS2seq/compare/v0.6.1...v0.6.2">0.6.2</a> (2022-08-23)</h2>
<h3 id="bug-fixes_1">Bug Fixes</h3>
<ul>
<li>change the directory with nextclade dataset from temp to non-temp in order to circumvent a WorkflowError (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/ae3da1169da76358a156e7e9befa5790f2261d35">ae3da11</a>)</li>
</ul>
<h2 id="061-2022-08-15"><a href="https://github.com/RIVM-bioinformatics/SARS2seq/compare/v0.6.0...v0.6.1">0.6.1</a> (2022-08-15)</h2>
<h3 id="bug-fixes_2">Bug Fixes</h3>
<ul>
<li>write AmpliGone output to logfile instead of stdout (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/ff16fe4f6436e56d31410bac45ad13b4d6fefef3">ff16fe4</a>)</li>
</ul>
<h3 id="dependencies_5">Dependencies</h3>
<ul>
<li>simplify the <code>QC_and_cleanup</code> environment (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/8a8059ee2131988bb6432bb5fa4dbe5d823d239e">8a8059e</a>)</li>
<li>update Nextclade to newer (lenient) version 2.4.x (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/d5b90e2d452f3999417fb655746f6768326e432a">d5b90e2</a>)</li>
<li>Update required snakemake version to 7.x (pinned to 7.12.1) (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/b171fc4c9dcc4882bb4e827ac73fd4b5c682389f">b171fc4</a>)</li>
<li>upgrade AmpliGone to version 1.1.0 (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/ec005aa58f3d1a98ec45ed7f4ab669f27ef1a055">ec005aa</a>)</li>
<li>use the bioconda channel to install AmpliGone instead of pip (<a href="https://github.com/RIVM-bioinformatics/SARS2seq/commit/8a8059ee2131988bb6432bb5fa4dbe5d823d239e">8a8059e</a>)</li>
</ul>
<h2 id="060-2022-04-26"><a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/compare/v0.5.4...v0.6.0">0.6.0</a> (2022-04-26)</h2>
<h3 id="features">Features</h3>
<ul>
<li>Update TrueConsense to 0.5.0 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/972c08a3fd88e8ade9c6095ec84b3645306dcf05">972c08a</a>)</li>
</ul>
<h3 id="bug-fixes_3">Bug Fixes</h3>
<ul>
<li>Update Ampligone to 1.0.1 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/cce0e31fa8b5a82a368936c526d2935d58620af6">cce0e31</a>)</li>
</ul>
<h3 id="054-2022-04-08"><a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/compare/v0.5.3...v0.5.4">0.5.4</a> (2022-04-08)</h3>
<h3 id="dependencies_6">Dependencies</h3>
<ul>
<li>pin pangolin version number to 4.0.2 without auto-updating (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/27e30d34940d852993f616b04feae231751bf607">27e30d3</a>)</li>
</ul>
<h3 id="053-2022-04-05"><a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/compare/v0.5.2...v0.5.3">0.5.3</a> (2022-04-05)</h3>
<h3 id="bug-fixes_4">Bug Fixes</h3>
<ul>
<li>allow pangolin v4 empty output when there's low coverage in the typing sequence (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/32372022b5f35a1d4332d4ad72eabc258b5ec2ea">3237202</a>)</li>
</ul>
<h3 id="052-2022-04-04"><a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/compare/v0.5.1...v0.5.2">0.5.2</a> (2022-04-04)</h3>
<h3 id="dependencies_7">Dependencies</h3>
<ul>
<li>add python3.7 to Alignment env to stabilize environment during installation (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/6f1387d9a10dbbcffe9dc664b7c160187288136b">6f1387d</a>)</li>
<li>change Typing rules and packages for compatibility with pangolin v4.* (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/81a61d55ea320c516aaf1d6972638b22af8b3321">81a61d5</a>)</li>
</ul>
<h3 id="051-2022-02-04"><a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/compare/v0.5.0...v0.5.1">0.5.1</a> (2022-02-04)</h3>
<h3 id="bug-fixes_5">Bug Fixes</h3>
<ul>
<li>finish an analysis when empty input fastq files are given (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/7a30b18e1b149ac498cdd53902c9d1f2e258b1f7">7a30b18</a>)</li>
</ul>
<h3 id="dependencies_8">Dependencies</h3>
<ul>
<li>update AmpliGone 0.4.0 --&gt; 0.4.1 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/cf64dc459be28e0b74abd1c2207d00e950b788c9">cf64dc4</a>)</li>
<li>update AmpliGone 0.4.1 --&gt; 0.4.3 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/28baf2a8becdc0e3945347a2483561fd119d728b">28baf2a</a>)</li>
</ul>
<h2 id="050-2022-01-27"><a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/compare/v0.4.2...v0.5.0">0.5.0</a> (2022-01-27)</h2>
<h3 id="features_1">Features</h3>
<ul>
<li>Write a short report with used/configured settings of the analysis (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/ffe3141c2b3af151a108a2d768584e6d30cc1557">ffe3141</a>)</li>
</ul>
<h3 id="bug-fixes_6">Bug Fixes</h3>
<ul>
<li>properly exit with right exit codes (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/4ec630723960f41f47c53e668d89f0f8a2cf1f8f">4ec6307</a>)</li>
<li>sucessfully finish an analysis when no primers are given (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/491673192ea83293fb2fe639e833e7d712ea311e">4916731</a>)</li>
</ul>
<h3 id="performance-improvements">Performance Improvements</h3>
<ul>
<li>tweak read qc filtering settings (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/27916d4a9dc6be9b402223eed36654886b55b541">27916d4</a>)</li>
</ul>
<h3 id="dependencies_9">Dependencies</h3>
<ul>
<li>add FPDF version 1.7.2 to dependency list(s) (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/1a192ecc146bd2d6eb1f07c0ea3f12331a042cf1">1a192ec</a>)</li>
<li>add urllib3 to dependency list for the auto-updater (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/f3b9b536228dd077a77e56d414b394c161ece147">f3b9b53</a>)</li>
<li>include conda (lenient) version 4.11.x in main environment to circumvent python3.10 bug (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/a7821f463dd6157bea2e109f17cab9002cd138e1">a7821f4</a>)</li>
<li>remove pysamstats from 'clean' environment (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/485c94ed732057a572f6a3e631aa5618f286e9cf">485c94e</a>)</li>
<li>update (pangolin) snakemake 6.4.1 --&gt; 6.8.0 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/1ffc79dbb1669cf5df06b997ba28786690e9bde3">1ffc79d</a>)</li>
<li>update AmpliGone version v0.3.3 --&gt; v0.4.0 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/999b757518e40d6ea10fbd31ef35d7da81ff8772">999b757</a>)</li>
<li>update bcftools 1.12 --&gt; 1.14 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/91080763fdfb65d4bb90372c7cef7d8302606f7b">9108076</a>)</li>
<li>update bedtools 2.29.2 --&gt; 2.30.0 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/91080763fdfb65d4bb90372c7cef7d8302606f7b">9108076</a>)</li>
<li>update biopython 1.78 --&gt; 1.79 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/91080763fdfb65d4bb90372c7cef7d8302606f7b">9108076</a>)</li>
<li>update biopython to version 1.79 in main environment (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/43251f09945d78cd3a7952d24256e8c703a915f9">43251f0</a>)</li>
<li>update FastP to version 0.23.2 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/b134b64644dbdb3af93c1bba3d23d3d50660e557">b134b64</a>)</li>
<li>update fastqc 0.11.8 --&gt; 0.11.9 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/485c94ed732057a572f6a3e631aa5618f286e9cf">485c94e</a>)</li>
<li>update minimap2 &amp; mappy 2.17 --&gt; 2.24 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/cb99be4361fbe232be214f0c1c556e560bf76201">cb99be4</a>)</li>
<li>update multiqc 1.9 --&gt; 1.11 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/485c94ed732057a572f6a3e631aa5618f286e9cf">485c94e</a>)</li>
<li>update nextclade 1.6.0 --&gt; 1.9 + change to lenient versioning (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/1ffc79dbb1669cf5df06b997ba28786690e9bde3">1ffc79d</a>)</li>
<li>update nextclade lenient version v1.9.x --&gt; v1.10.x (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/2e1cf9ef2ef8c81bbb6219852df929a2f0fd684a">2e1cf9e</a>)</li>
<li>update pandas 1.2.3 --&gt; 1.3.5 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/485c94ed732057a572f6a3e631aa5618f286e9cf">485c94e</a>)</li>
<li>update pangolin 3.0 --&gt; 3.1 + change to lenient versioning (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/1ffc79dbb1669cf5df06b997ba28786690e9bde3">1ffc79d</a>)</li>
<li>update parmap 1.5.2 --&gt; 1.5.3 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/485c94ed732057a572f6a3e631aa5618f286e9cf">485c94e</a>)</li>
<li>update PyYaml 5.4.1 --&gt; 6.0 and fix compatibility (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/e9a1ef955948ec30ecb6960a69e9da7d4ef8bd27">e9a1ef9</a>)</li>
<li>update samtools 1.10 --&gt; 1.14 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/cb99be4361fbe232be214f0c1c556e560bf76201">cb99be4</a>)</li>
<li>update seqkit 0.14.0 --&gt; 2.1.0 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/cb99be4361fbe232be214f0c1c556e560bf76201">cb99be4</a>)</li>
<li>update snakemake to version 6.13.1 in main environment (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/43251f09945d78cd3a7952d24256e8c703a915f9">43251f0</a>)</li>
<li>update tqdm 4.59 --&gt; 4.62 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/485c94ed732057a572f6a3e631aa5618f286e9cf">485c94e</a>)</li>
</ul>
<h3 id="documentation_1">Documentation</h3>
<ul>
<li>add extra primer keywords (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/422f19238d4f91155bd9e6aafba3a8c1dd56aa6a">422f192</a>)</li>
<li>fix typo in docs and update download instructions (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/9f239e3bf778f2df103a2fc51b0343cd68261b8d">9f239e3</a>)</li>
<li>update readme with correct primer-keywords (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/17878f84d18774e3e0408dcf13bc555b45b3171c">17878f8</a>)</li>
<li>update snakemake version label (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/f1a474df2cb7556768aabd839a2d9e22910719cd">f1a474d</a>)</li>
</ul>
<h3 id="042-2021-12-08"><a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/compare/v0.4.1...v0.4.2">0.4.2</a> (2021-12-08)</h3>
<h3 id="dependencies_10">Dependencies</h3>
<ul>
<li>remove manual installation of pango-designations repo (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/5527d53f554203c5ac59d45df124247ab6eb5396">5527d53</a>)</li>
<li>update AmpliGone to version 0.3.3 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/b97a32ea64054d1581761a9c0e82d0f31b89039e">b97a32e</a>)</li>
<li>update nextclade typingtool to version 1.6.0 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/303d4d9ebc7a6fff966ef4f7a4bd1c1a130c84f8">303d4d9</a>)</li>
</ul>
<h3 id="041-2021-11-26"><a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/compare/v0.4.0...v0.4.1">0.4.1</a> (2021-11-26)</h3>
<h3 id="dependencies_11">Dependencies</h3>
<ul>
<li>update AmpliGone to version 0.3.2 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/d47f62bd7c8eb632e1f492492efcc76a95ff0ea4">d47f62b</a>)</li>
</ul>
<h2 id="040-2021-11-25"><a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/compare/v0.3.0...v0.4.0">0.4.0</a> (2021-11-25)</h2>
<h3 id="features_2">Features</h3>
<ul>
<li>allow SARS2seq to update itself to latest release (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/41b711b6de543d5a6bae2ec6aa10a8afb2ad4917">41b711b</a>)</li>
<li>override consensus-index with overlap coordinates index (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/b81fd1cf16e95bc113f5598b3094b0262cce1b95">b81fd1c</a>)</li>
</ul>
<h3 id="bug-fixes_7">Bug Fixes</h3>
<ul>
<li>Remove --end-bonus and -k28, update -A4 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/e7f8bf7000f223761df736a83f2bf4b18c81d60c">e7f8bf7</a>)</li>
<li>update sars-cov-2 genomic features GFF to include missing ORF (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/4399cf4ba6ef45a68225dad2c1139b74348bf1cc">4399cf4</a>)</li>
</ul>
<h3 id="performance-improvements_1">Performance Improvements</h3>
<ul>
<li>Update minimap2 settings for nanopore data (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/75bb47546a4a108a003c7e694df9ba58972a80bc">75bb475</a>)</li>
</ul>
<h3 id="documentation_2">Documentation</h3>
<ul>
<li>add (auto-)updating behaviour and "skip-updates" flag to documentation (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/43c36558dfd9f7dd1cf910005ae91facd8c2c439">43c3655</a>)</li>
</ul>
<h3 id="dependencies_12">Dependencies</h3>
<ul>
<li>pin TrueConsense to version 0.3.0 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/b6dfd46bf8ad56b87dca6a528efec51d733172a3">b6dfd46</a>)</li>
<li>update AmpliGone to version 0.3.0 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/f0284030361b5d0bcf0153ea02715efe758c5b35">f028403</a>)</li>
<li>update AmpliGone to version 0.3.1 (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/29f15168fd21c69b7ccdc8c1917da53f10da0253">29f1516</a>)</li>
</ul>
<h3 id="styles">Styles</h3>
<ul>
<li>remove trailing space (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/9933ab4c7d27535f6688bc39fc164ebc0eade6d2">9933ab4</a>)</li>
</ul>
<h2 id="030-2021-11-09"><a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/compare/v0.2.2...v0.3.0">0.3.0</a> (2021-11-09)</h2>
<h3 id="features_3">Features</h3>
<ul>
<li>flexible memory requirements per workflow job (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/dd1c89e2ddf6f3b7bec0ca48452aabce69da2942">dd1c89e</a>)</li>
</ul>
<h3 id="bug-fixes_8">Bug Fixes</h3>
<ul>
<li>use 'map-ont' mm2 preset in nanopore workflow (<a href="https://www.github.com/RIVM-bioinformatics/SARS2seq/commit/87e056c03329638ec95e00533e2e16d80424ae52">87e056c</a>)</li>
</ul></div>
        </div>

    <footer class="col-md-12">
        <hr>
        
        <center>copyright Â© 2021 RIVM</center>
        
    </footer>

        <script src="../js/jquery.min.js"></script>
        <script src="../js/bootstrap.min.js"></script>
        <script src="../js/highlight.min.js"></script>
        <script>hljs.initHighlightingOnLoad();</script>
        <script src="../js/base.js"></script>
        <script src="../search/main.js"></script>
    </body>
</html>